server/service/mdm.go

```diff
@@ -1980,3 +1980,138 @@ func (svc *Service) UpdateMDMDiskEncryption(ctx context.Context, teamID *uint, e
+	if err := svc.authz.Authorize(ctx, &fleet.MDMConfigProfileAuthz{TeamID: host.TeamID}, fleet.ActionWrite); err != nil {
+		return ctxerr.Wrap(ctx, err)
+	}
+
+	// TODO: prescreen host platform against profile prefix?
+
+	var profileTeamID *uint
+	var profileName string
+	switch {
+	case strings.HasPrefix(profileUUID, fleet.MDMAppleProfileUUIDPrefix), strings.HasPrefix(profileUUID, fleet.MDMAppleDeclarationUUIDPrefix):
+		if err := svc.VerifyMDMAppleConfigured(ctx); err != nil {
+			return ctxerr.Wrap(ctx, fleet.NewInvalidArgumentError("profile_uuid", fleet.AppleMDMNotConfiguredMessage).WithStatus(http.StatusBadRequest), "check apple mdm enabled")
+		}
+		prof, err := svc.ds.GetMDMAppleConfigProfile(ctx, profileUUID)
+		if err != nil {
+			return ctxerr.Wrap(ctx, err, "getting apple config profile")
+		}
+		profileTeamID = prof.TeamID
+		profileName = prof.Name
+
+	case strings.HasPrefix(profileUUID, fleet.MDMAppleDeclarationUUIDPrefix):
+		if err := svc.VerifyMDMAppleConfigured(ctx); err != nil {
+			return ctxerr.Wrap(ctx, fleet.NewInvalidArgumentError("profile_uuid", fleet.AppleMDMNotConfiguredMessage).WithStatus(http.StatusBadRequest), "check apple mdm enabled")
+		}
+		decl, err := svc.ds.GetMDMAppleDeclaration(ctx, profileUUID)
+		if err != nil {
+			return ctxerr.Wrap(ctx, err, "getting apple declaration")
+		}
+		profileTeamID = decl.TeamID
+		profileName = decl.Name
+
+	case strings.HasPrefix(profileUUID, fleet.MDMWindowsProfileUUIDPrefix):
+		if err := svc.VerifyMDMWindowsConfigured(ctx); err != nil {
+			return ctxerr.Wrap(ctx, fleet.NewInvalidArgumentError("profile_uuid", fleet.WindowsMDMNotConfiguredMessage).WithStatus(http.StatusBadRequest), "check windows mdm enabled")
+		}
+		prof, err := svc.ds.GetMDMWindowsConfigProfile(ctx, profileUUID)
+		if err != nil {
+			return ctxerr.Wrap(ctx, err, "getting windows config profile")
+		}
+		profileTeamID = prof.TeamID
+		profileName = prof.Name
+
+	default:
+		// TODO: confirm error message?
+		return ctxerr.Wrap(ctx, fleet.NewInvalidArgumentError("profile_uuid", "Invalid profile UUID prefix."))
+	}
+
+	// check again based on team id of profile before we proceeding
+	if err := svc.authz.Authorize(ctx, &fleet.MDMConfigProfileAuthz{TeamID: profileTeamID}, fleet.ActionWrite); err != nil {
+		return ctxerr.Wrap(ctx, err)
+	}
+
+	// TODO: specs say to screen disk encryption profile? what about other reserved profiles?
+	reservedNames := mdm.FleetReservedProfileNames()
+	if _, ok := reservedNames[profileName]; ok {
+		return fleet.NewInvalidArgumentError("profile_uuid", fmt.Sprintf("%s profile cannot be resent.", profileName))
+	}
+
+	var hTID, pTID uint
+	if host.TeamID != nil {
+		hTID = *host.TeamID
+	}
+	if profileTeamID != nil {
+		pTID = *profileTeamID
+	}
+	if hTID != pTID {
+		// TODO: confirm error message? should this be an auth error?
+		return ctxerr.Wrap(ctx, fleet.NewInvalidArgumentError("profile_uuid", "Profile does not belong to host's team."))
+	}
```

Comment:
in general for all the team id checks, I'm thinking, what if this is already covered by `GetHostMDMProfileStatus`?

in summary:

- Check that user can read the host (done)
- Check that the user can modify the host team profiles (done)

Then, if the host has assigned the profile, it's fair game. We do this  check via `GetHostMDMProfileStatus` already.

Reply:
@roperzh, that makes sense, and I had considered it as a way to simplify. My concern was that it made the auth reliant on the implicit assumption that the host mdm profiles table would only have a profile associated with a host where the host and the profile belong to the same team. Of course, that's how the table and the crons are designed to work today. One case where that isn't necessarily true is when a host switches teams and the table still includes profiles being removed. We're currently saying that "pending" status (which would include pending removals) can't be resent so it feels pretty safe to make all these assumptions now. But it is pretty subtle how the pieces fit together and something that might not be obvious to folks in the future if things change. 

At the end of the day, I think we want to ensure that a user to be able to resend a profile that doesn't belong to the same team as the host. I tried to be explicit about checking that here because of the subtleties. I don't mind changing the approach, but I figured auth was one of those places where it pays to be explicit and this approach was the most explicit way I found (at the potential expense of being redundant and verbose).

Reply:
my 2cents but I'm ok with doing that check as well is that all this logic is optimizing for a very unlikely edge case that, might even be OK. 

what I'm proposing is: if the host has the profile at the moment the API is hit, it's okay to re-send it.

once again, fine with this approach too. In that case there are a few things we can optimize, just lmk and I'll do another review with that in mind 🚀

Reply:
Sounds good to me